kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt

```diff
@@ -206,12 +205,12 @@ internal abstract class AbstractSendChannel<E> : SendChannel<E> {
      * * ENQUEUE_FAILED -- buffer is not full (should not enqueue)
      * * ReceiveOrClosed<*> -- receiver is waiting or it is closed (should not enqueue)
      */
-    private fun enqueueSend(send: SendElement): Any? {
+    private fun enqueueSend(send: Send): Any? {
         if (isBufferAlwaysFull) {
-            queue.addLastIfPrev(send, { prev ->
+            queue.addLastIfPrev(send) { prev ->
                 if (prev is ReceiveOrClosed<*>) return@enqueueSend prev
                 true
-            })
+            }
```

Comment:
Line 220: `return ENQUEUE_FAILED`

Nobody checks for `ENQUEUE_FAILED`. It doesn't look like a bug because of the current state machine (`null` -> success, `Closed` -> closed, everything else -> retry), but kind of error-prone.

For example, `null` can become `ENQUEUE_SUCCESS` and `ENQUEUE_FAILED` will become `null` instead. Code flow will be more straightforward and explicit

Reply:
I've rewritten the corresponding `when` statements in a more explicit way, checking all possible outcomes and added `error` on unexpected result.